### ISSUE TITLE
Speed issue

### ISSUE THREAD

**Issue by mrsanvicente** [jookies/jasmin#648]
Hello,

When i send more that 2 sms  the smpp client gets:  (SMPPs Throughput is set to 100 or any higher number)
Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)", Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)", Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)", Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)"

 ERROR    1491 QoS: submit_sm_event is faster (0:00:00.002545) than fixed throughput 

But when i set SMPPs Throughput to 0: i got the following error

Submit_sm - resp: "Unknown error",
 CRITICAL 1491 Got an unknown exception: float division by zero

Are you aware of this bug??

**Comment by farirat** [comment]
hey @mrsanvicente 

This is a strange issue:

1. Which version ?
2. Please clarify if this is about smppc traffic (upstream) or smpps traffic (downstream) ?
3. Please copy/paste your user config (user -s, if the issue on downstream) or your smpp connector (smppccm -s, if the issue on upstream), take care of anonymizing your sensitive data

Thanks !

**Comment by mrsanvicente** [comment]
Hello @farirat 

1.It should be the latest.  how can i verify
2.Downstream  
3.
jcli : user -s xxx
username xxx
mt_messaging_cred defaultvalue src_addr 444444
mt_messaging_cred quota http_throughput 100.0
mt_messaging_cred quota balance 99.981
mt_messaging_cred quota smpps_throughput 110.0
mt_messaging_cred quota sms_count 81
mt_messaging_cred quota early_percent ND
mt_messaging_cred valuefilter priority ^[0-3]$
mt_messaging_cred valuefilter content .*
mt_messaging_cred valuefilter src_addr .*
mt_messaging_cred valuefilter dst_addr .*
mt_messaging_cred valuefilter validity_period ^\d+$
mt_messaging_cred authorization http_send True
mt_messaging_cred authorization http_dlr_method True
mt_messaging_cred authorization hex_content True
mt_messaging_cred authorization http_balance True
mt_messaging_cred authorization smpps_send True
mt_messaging_cred authorization priority True
mt_messaging_cred authorization http_long_content True
mt_messaging_cred authorization src_addr True
mt_messaging_cred authorization dlr_level True
mt_messaging_cred authorization http_rate True
mt_messaging_cred authorization validity_period True
mt_messaging_cred authorization http_bulk False
uid qual1tel
smpps_cred quota max_bindings ND
smpps_cred authorization bind True
gid test



ERROR, no matter the value on  smpps_throughput.:
Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)",

NOTE      
if: smpps_throughput=0  then i get 
Submit_sm - resp: "Unknown error"

NOTE: in order to make it work i have to set TPS=1 on the smpp client.

Thanks

**Comment by tarikogut** [comment]
change to submit_Throughput  110 to 80 test again

2018-01-20 0:39 GMT+03:00 mrsanvicente <notifications@github.com>:

**Comment by mrsanvicente** [comment]
Hello tarikogut,

I tested with smpp throuput to 80 and many others.  The only one working is TPS=1.

Error:  SMPP Submit_sm - resp: "Throttling error (ESME exceeded allowed message limits)"



Thanks

**Comment by tarikogut** [comment]
which operator using now ?? which one ??

2018-01-21 9:31 GMT+03:00 mrsanvicente <notifications@github.com>:

**Comment by farirat** [comment]
@mrsanvicente just made a check, cant reproduce your issue.

Please check your default-smpps_01.log and router.log, i guess you're having some quota limitation.

**Comment by mrsanvicente** [comment]
Hi

@tarikogut , it is a internal setup, before send it to the operator.

@farirat   I find the following log,  

2018-01-21 15:05:34 ERROR    1491 QoS: submit_sm_event is faster (0:00:00.002500) than fixed throughput (0:00:00.012500) for user (fsdtel), rejecting message.

Any idea  how to fix that?

Regards

**Comment by farirat** [comment]
@mrsanvicente no elements to reproduce = no bug :)

**Comment by SahinHaberlesme** [comment]
ESME Service Limited 1 connection :) this not bug this is support :)

**Comment by mrsanvicente** [comment]
Hi  @farirat @SahinHaberlesme ,

OK there is no bug,   but i cant find where is the limitation.  **The issue is with any user, so i looks like a general config.  Can you help me?**

Log:  2018-01-21 15:05:34 ERROR 1491 QoS: submit_sm_event is faster (0:00:00.002500) than fixed throughput (0:00:00.012500) for user (fsdtel), rejecting message.


Config

jcli : user -s fsdtel
username fsdtel
mt_messaging_cred defaultvalue src_addr 1144400016
mt_messaging_cred quota http_throughput 100.0
mt_messaging_cred quota balance 99.473
mt_messaging_cred quota smpps_throughput 80.0
mt_messaging_cred quota sms_count 71
mt_messaging_cred quota early_percent 100.0
mt_messaging_cred valuefilter priority ^[0-3]$
mt_messaging_cred valuefilter content .*
mt_messaging_cred valuefilter src_addr .*
mt_messaging_cred valuefilter dst_addr .*
mt_messaging_cred valuefilter validity_period ^\d+$
mt_messaging_cred authorization http_send True
mt_messaging_cred authorization http_dlr_method True
mt_messaging_cred authorization hex_content True
mt_messaging_cred authorization http_balance True
mt_messaging_cred authorization smpps_send True
mt_messaging_cred authorization priority True
mt_messaging_cred authorization http_long_content True
mt_messaging_cred authorization src_addr True
mt_messaging_cred authorization dlr_level True
mt_messaging_cred authorization http_rate True
mt_messaging_cred authorization validity_period True
mt_messaging_cred authorization http_bulk False
uid fsdtel
smpps_cred quota max_bindings 4
smpps_cred authorization bind True
gid test

**Comment by sah-anshu** [comment]
@mrsanvicente
Please answer few of my queries:

1. Server Load & Free RAM
2. Are you running any Interceptors.

Cheers!

**Comment by mrsanvicente** [comment]
Hello @sah-anshu 

1. NO LOAD,   Free RAM 7G of 8G
2.  Yes i have an interceptor which does:  A number validation, Sets ton,  B number lookup in a Portability DB.

Thanks

**Comment by sah-anshu** [comment]
@mrsanvicente thanks for update.
Now, 
1. This NUMBER Validation API is Local or Using any 3rd party service ?
2. Portability Lookup is Local or remote ?
3. If this is remote services, can you evaluate how much time this process is consuming ?

Thanks

**Comment by mrsanvicente** [comment]
@sah-anshu 

1. The number validation is Local.
2.The Portability lookup is with a local script, but it sends a remote redis query.
3. I can check the speed and send back to you the timings.    If so is the isse,   What is your recommendations to achieve high throughput with all this validations??

**Comment by farirat** [comment]
The interceptor script is run synchronously, this means that is your call to an external DB have some consequent delay (the DB may be overloaded, or too slow) and this will impact your messaging flow, you need to care about your third party systems performance if you are relying on them in interception.

This behavior makes the QoS limitation logic, if your DB is locking randomly on some queries it is replying a bunch of them in same time, after interceptor there's the router where the QoS is measured, the router gets fooled by your DB performance and blocks a lot of messages.

In this situation, i suggest you one of these possibilities:

1. Bring a telecom-grade DB that can respond quick, apply some caching on top of it to avoid multiple queries for same number in a defined timelapse,
2. Remove the QoS control from the router and implement one yourself inside your interceptor.

**Comment by mrsanvicente** [comment]
@farirat 
Thanks for your inputs.
1. I think that i can upgrade the script to do the local query. And do caching.  I will get back to you with results.
2. How can i remove the QoS control from the router and implement it inside the interceptor??

I also have another question.

If the interceptor works synchronally.     But the smpp user sends   asynchronally.     Will the router still reject the MTs with the same error  ("Throttling error (ESME exceeded allowed message limits)"?      Can the messages be queued???

Thank you

**Comment by sah-anshu** [comment]
@mrsanvicente how is the transaction time without Interceptor?

**Comment by mrsanvicente** [comment]
@sah-anshu 

Here my results..
-not using an interceptor but enabled on the system

	a) Sync  1 sessions,   1tps, ok 
	b) Sync  10 sessions,   13tps, ok 	Max,  more tps fails
   
	c) Async  not working, not even 2tps

	 SMPP Submit_sm - resp: "Ok", Submit_sm - resp: "Throttling error

**Comment by mrsanvicente** [comment]
Hello @sah-anshu @farirat 

When i disabled the interceptor i can reach higher speed.  I tested up to 5000tps

-Interceptor Not enabled

 	a)Sync 1 session,  13tps, ok

	Async
        110 tps * 10 segs, ok
       1000 tps * 10 segs, ok
       5000tps * 1 segs, ok

I also tested that jasmin can receive asynch sms and then send sync, buffering the sms.  My solution for handlings higher speed with interceptors would be to receive in one jasmin node, and then load balance between other jasmin nodes with interceptor enabled.

Thanks for your inputs!
